internal/lsp/watcher/global_watcher.go

```diff
@@ -0,0 +1,450 @@
+	}
+
+	return nil
+}
+
+// shouldIgnoreDirectory checks if a directory should be ignored based on hierarchical
+// .gitignore/.crushignore files. It checks for ignore files in each directory from
+// the target directory up to the workspace root.
+func (gw *globalWatcher) shouldIgnoreDirectory(dirPath string) bool {
+	if gw.workspaceRoot == "" {
+		return false
+	}
+
+	if strings.HasPrefix(dirPath, gw.workspaceRoot) {
+		return gw.isIgnoredInWorkspace(dirPath, gw.workspaceRoot)
+	}
+
+	return false
+}
+
+// isIgnoredInWorkspace checks if a path is ignored by walking up the directory tree
+// from the target path to the workspace root, checking for .gitignore/.crushignore
+// files in each directory.
+func (gw *globalWatcher) isIgnoredInWorkspace(targetPath, workspaceRoot string) bool {
+	// Get relative path from workspace root
+	relPath, err := filepath.Rel(workspaceRoot, targetPath)
+	if err != nil {
+		return false
+	}
+
+	// Don't ignore the workspace root itself
+	if relPath == "." {
+		return false
+	}
+
+	// Check each directory from workspace root to target path for ignore files
+	currentPath := workspaceRoot
+	pathComponents := strings.Split(relPath, string(filepath.Separator))
+
+	for i, component := range pathComponents {
+		currentPath = filepath.Join(currentPath, component)
+
+		// Build the relative path from workspace root to current path
+		currentRelPath := strings.Join(pathComponents[:i+1], "/")
+
+		// Check ignore files in each parent directory
+		checkPath := workspaceRoot
+		for j := 0; j <= i; j++ {
+			if j > 0 {
+				checkPath = filepath.Join(checkPath, pathComponents[j-1])
+			}
+
+			// Check .gitignore in this directory
+			if gw.checkIgnoreFile(filepath.Join(checkPath, ".gitignore"), currentRelPath) {
+				return true
+			}
+
+			// Check .crushignore in this directory
+			if gw.checkIgnoreFile(filepath.Join(checkPath, ".crushignore"), currentRelPath) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// checkIgnoreFile checks if a path matches patterns in an ignore file
+func (gw *globalWatcher) checkIgnoreFile(ignoreFilePath, relPath string) bool {
+	content, err := os.ReadFile(ignoreFilePath)
+	if err != nil {
+		return false // File doesn't exist or can't be read
+	}
+
+	lines := strings.Split(string(content), "\n")
+	ignoreParser := ignore.CompileIgnoreLines(lines...)
+
+	// Check both with and without trailing slash for directories
+	if ignoreParser.MatchesPath(relPath) {
+		return true
+	}
+
+	// For directories, also check with trailing slash
+	if ignoreParser.MatchesPath(relPath + "/") {
+		return true
+	}
+
+	return false
+}
+
+// addDirectoryToWatcher adds a directory to the fsnotify watcher.
+// fsnotify handles deduplication internally, so we don't need to track watched directories.
+func (gw *globalWatcher) addDirectoryToWatcher(dirPath string) error {
+	if gw.watcher == nil {
+		return fmt.Errorf("global watcher not initialized")
```

Comment:
The same apply to errors. Actually, for errors it's even more important (and a kinda of Go convention, even).

```suggestion
		return fmt.Errorf("lsp watcher: global watcher not initialized")
```